### PR TITLE
status: implement using the new channel-map endpoint

### PR DESCRIPTION
**Pull request by sergiusens**
This enables progressive release reporting from status.

- [ ] Have you followed the [guidelines for contributing](https://github.com/snapcore/snapcraft/blob/master/CONTRIBUTING.md)?
- [ ] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?
- [ ] Have you successfully run `./runtests.sh static`?
- [ ] Have you successfully run `./runtests.sh tests/unit`?

-----

### CODE REVIEW

File: snapcraft/cli/store.py [hunk 1]
```diff
@@ -392,34 +393,31 @@ def close(snap_name, channels):
...
+        else:
+            architectures = existing_architectures
+        click.echo(get_tabulated_channel_map(snap_channel_map, architectures))
 
 
 @storecli.command("list-revisions")
 @click.option(
     "--arch", metavar="<arch>", help="The snap architecture to get the status for"
 )
-@click.option(
```
Comment (sergiusens) [inline snapcraft/cli/store.py]: oops

File: snapcraft/storeapi/_snap_v2_client.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+            url,
+            headers={
+                "Authorization": auth,
+                "Content-Type": "application/json",
+                "Accept": "application/json",
+            },
+        )
+
+        if not response.ok:
+            raise errors.StoreSnapChannelMapError(snap_name=snap_name)
```
Comment (cjp256) [inline snapcraft/storeapi/_snap_v2_client.py]: Is there a way to include the response error, if available?

File: snapcraft/storeapi/v2/snap_channel_map.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+            ):
+                return channel_mapping
+        raise ValueError(
+            f"No channel mapped to {channel_name!r} for architecture {architecture!r}"
+        )
+
+    def get_channel_info(self, channel_name: str) -> channel.SnapChannel:
+        for snap_channel in self.snap.channels:
+            if snap_channel.name == channel_name:
+                return snap_channel
```
Comment (cjp256) [inline snapcraft/storeapi/v2/snap_channel_map.py]: I'm shocked that you're not using list comprehension (or generators) here :joy:  lmftfy
```
try:
  return next(x for x in self.snap.channels if x.name == channel_name)
except StopIteration:
  return ValueError(..)
```

But seriously, I assume all of these ValueError() cases should never happen?

File: snapcraft/cli/_channel_map.py [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+from typing import List, Union, TYPE_CHECKING
+
+from tabulate import tabulate
+
+if TYPE_CHECKING:
+    from snapcraft.storeapi.v2 import snap_channel_map
+
+
+def _get_channel_hint(*, channel_map, fallback: str, architecture: str) -> str:
+    tick = "^"
```
Comment (cjp256) [inline snapcraft/cli/_channel_map.py]: Any chance you can use a constant/enum for these tick types?

File: snapcraft/cli/_channel_map.py [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+                else:
+                    architecture_string = ""
+
+                channel_lines.append(
+                    [track_string, architecture_string]
+                    + _get_channel_line(snap_channel_map, channel_name, architecture)
+                )
+
+    # Item 6 is expiration_date
+    have_expiration = any(line[6] != "" for line in channel_lines)
```
Comment (cjp256) [inline snapcraft/cli/_channel_map.py]: Can you show an example in the comments?

File: snapcraft/storeapi/v2/snap_channel_map.py [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+
+class SnapChannelMap:
+    """Represents the data returned from the channel-map call from the Snap Store."""
+
+    def __repr__(self) -> str:
+        return "<SnapChannelMap: {!r}>".format(self.snap.name)
+
+    def __init__(self, snap_channel_map: Dict[str, Any]) -> None:
+        self._payload = snap_channel_map
+        self._channelmap: Optional[List[channel.MappedChannel]] = None
```
[the remark this answers is not in the document]
Comment (sergiusens) [reply snapcraft/storeapi/v2/snap_channel_map.py]: ah, this is a result of the API changing from channelmap to channel-map mid development, I just missed this one!

File: snapcraft/storeapi/_snap_v2_client.py (hunk 2 above)
Comment (sergiusens) [reply snapcraft/storeapi/_snap_v2_client.py]: I would need to ask Natalia about what the possible error codes are for this.

File: snapcraft/storeapi/v2/revision.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+
+class Revision:
+    def __repr__(self) -> str:
+        return "<Revision: {!r} for version {!r} and architectures {!r}>".format(
+            self.revision, self.version, self.architectures
+        )
+
+    def __init__(self, revision: Dict[str, Any]) -> None:
+        self._payload = revision
```
[the remark this answers is not in the document]
Comment (sergiusens) [reply snapcraft/storeapi/v2/revision.py]: well, the rest of this module (snapcraft.storeapi) is not using this loading strategy (classmethod), so for consistency sake I would not do it in this PR at least. I did think about it, but prefer consistency.

File: snapcraft/storeapi/v2/snap_channel_map.py (hunk 3 above)
Comment (sergiusens) [reply snapcraft/storeapi/v2/snap_channel_map.py]: I just copied the structure of get_mapped_channel

ValueError here is a serious API error if it occurs so a good failure which is reported is good in my book.

File: snapcraft/storeapi/v2/snap_channel_map.py [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+        raise ValueError(f"No channel information for {channel_name!r}")
+
+    def get_revision(self, revision_number: int) -> revision.Revision:
+        for revision_item in self.revisions:
+            if revision_item.revision == revision_number:
+                return revision_item
+        raise ValueError(f"No revision information for {revision_number!r}")
+
+    def get_existing_architectures(self) -> Set[str]:
+        architectures: List[str] = list()
```
[the remark this answers is not in the document]
Comment (sergiusens) [reply snapcraft/storeapi/v2/snap_channel_map.py]: revision.architectures holds a list of strings, which is why I use "extend" bellow and not "append".

File: snapcraft/cli/_channel_map.py (hunk 4 above)
Comment (sergiusens) [reply snapcraft/cli/_channel_map.py]: yes I can!

**Pull request closed without merging**